File: Sinuka/Core/Repositories/UserRepository.py
from ..Factories.UserFactory import UserFactory
from ..Models.User import User
import bcrypt
import pymongo


class UserRepository:
    user_store: pymongo.collection.Collection = None

    def __init__(self, user_store: pymongo.collection.Collection):
        self.user_store = user_store

    def check_credentials(self, username: str, password: str) -> bool:
        user = self.find_user_by_username(username)
        if user is None:
            return False
        return bcrypt.checkpw(password.encode(), user.hashedPassword.encode())

    def find_user_by_username(self, username: str):
        user = self.user_store.find_one({"username": username})
        return UserFactory.parse_user_dict(user) if user is not None else None

    def find_user_by_email(self, email: str):
        user = self.user_store.find_one({"email": email})
        return UserFactory.parse_user_dict(user) if user is not None else None

    def add_user(self, user: User):
        return self.user_store.insert_one(vars(user))

File: Sinuka/src/Sinuka.Infrastructure/Database/Repositories/UserRepository.cs
using System.Threading.Tasks;
using System.Linq;
using Sinuka.Core.Interfaces.Repositories;
using Sinuka.Core.Interfaces.Services;
using Sinuka.Core.Models;

namespace Sinuka.Infrastructure.Database.Repositories
{
    public class UserRepository : IUserRepository
    {
        private readonly SinukaDbContext _dbContext;
        private readonly IHashingService _hashingService;

        public UserRepository(SinukaDbContext dbContext, IHashingService hashingService)
        {
            this._dbContext = dbContext;
            this._hashingService = hashingService;
        }

        public async Task AddUser(User newUser) => await this._dbContext.Users.AddAsync(newUser);

        public async Task<User?> CheckCredentials(string username, string password)
        {
            var user = await this.FindUserByUsername(username);
            if(user is null) return null;
            return this._hashingService.Compare(password, user.HashedPassword) ? user : null;
        }

        public async Task<User?> FindUserByEmail(string email) 
        {
            var result = await Task.FromResult(
                this._dbContext.Users.FirstOrDefault(
                    u => !u.DeletedAt.HasValue && u.Email.Email == email)
                    
            );
            return result;
        }

        public async Task<User?> FindUserByUsername(string username) 
        {
            var result = await Task.FromResult(
                this._dbContext.Users.FirstOrDefault(
                    u => !u.DeletedAt.HasValue && u.Username == username)
            );
            return result;
        }
    }
}
